engine/consensus/matching/engine_test.go

```diff
@@ -636,11 +641,198 @@ func (ms *MatchingSuite) TestRequestReceiptsPendingBlocks() {
+			if i < s {
+				ir.AddSignature(chunk.Index, unittest.IdentifierFixture(), unittest.SignatureFixture())
+			} else {
+
+				if i < n-int(ms.matching.approvalRequestsThreshold) {
+					expectedRequests = append(expectedRequests,
+						&messages.ApprovalRequest{
+							ResultID:   ir.Result.ID(),
+							ChunkIndex: chunk.Index,
+						})
+				}
+
+			}
+		}
+
+		ms.PendingResults[ir.ID()] = ir
+	}
+
+	exp := n - s - int(ms.matching.approvalRequestsThreshold)
+
+	// add an incorporated-result for a block that was already sealed. We
+	// expect that no approval requests will be sent for this result, even if it
+	// hasn't collected any approvals yet.
+	sealedBlockIR := unittest.IncorporatedResult.Fixture(
+		unittest.IncorporatedResult.WithResult(
+			unittest.ExecutionResultFixture(
+				unittest.WithBlock(&ms.LatestSealedBlock),
+			),
+		),
+		unittest.IncorporatedResult.WithIncorporatedBlockID(
+			ms.LatestSealedBlock.ID(),
+		),
+	)
+	ms.PendingResults[sealedBlockIR.ID()] = sealedBlockIR
+
+	// add an incorporated-result for an unfinalized block. It should not
+	// generate any requests either.
+	unfinalizedBlockIR := unittest.IncorporatedResult.Fixture(
+		unittest.IncorporatedResult.WithResult(
+			unittest.ExecutionResultFixture(
+				unittest.WithBlock(&unfinalizedBlock),
+			),
+		),
+		unittest.IncorporatedResult.WithIncorporatedBlockID(
+			unfinalizedBlock.ID(),
+		),
+	)
+	ms.PendingResults[unfinalizedBlock.ID()] = unfinalizedBlockIR
+
+	// wire-up the approval requests conduit to keep track of all sent requests
+	// and check that the target matches the unique verifier assigned to each
+	// chunk
+	requests := []*messages.ApprovalRequest{}
+	conduit := &mocknetwork.Conduit{}
+	conduit.On("Publish", mock.Anything, mock.Anything).
+		Return(nil).
+		Run(func(args mock.Arguments) {
+			// collect the request
+			ar, ok := args[0].(*messages.ApprovalRequest)
+			ms.Assert().True(ok)
+			requests = append(requests, ar)
+
+			// check that the target is the verifier assigned to the chunk
+			target, ok := args[1].(flow.Identifier)
+			ms.Assert().True(ok)
+			ms.Assert().Equal(verifier, target)
+		})
+	ms.matching.approvalConduit = conduit
+
+	err := ms.matching.requestPendingApprovals()
+	ms.Require().NoError(err)
+
+	// first time it goes through, no requests should be made because of the
+	// blackout period
+	ms.Assert().Len(requests, 0)
+
+	// Check the request tracker
+	ms.Assert().Equal(exp, len(ms.matching.requestTracker.index))
+	for _, expectedRequest := range expectedRequests {
+		requestItem := ms.matching.requestTracker.Get(
+			expectedRequest.ResultID,
+			expectedRequest.ChunkIndex,
+		)
+		ms.Assert().Equal(uint(0), requestItem.Requests)
+	}
+
+	// wait for the max blackout period to elapse and retry
+	time.Sleep(3 * time.Second)
+	err = ms.matching.requestPendingApprovals()
+	ms.Require().NoError(err)
+
+	// now we expect that requests have been sent for the chunks that haven't
+	// collected any approvals
```

Comment:
Nice. 

Do we have a test case to verify: if a result has 2 chunks, the consensus node has received an approval for the first chunk, then it will only fetch approval for the second chunk.

Reply:
Yes, the ExecutionResult fixture contains 17 chunks